src/vt_flash/state_idealidentity.jl

```diff
@@ -0,0 +1,199 @@
+
+    grad .-= grad2
+
+    # Scaling
+    @. grad[1:end-1] *= sqrt(nmol1 * nmol2 / nmolb)
+    grad[end] *= volumeb / sqrt(sum(nmolb))
+
+    return grad
+end
+
+function hessian!(
+    hess::AbstractMatrix,
+    state::IdealIdentityState,
+    mix::BrusilovskyEoSMixture,
+    nmolb::AbstractVector,
+    volumeb::Real,
+    RT::Real;
+    buf::BrusilovskyThermoBuffer=thermo_buffer(mix),
+)
+    # Here is version that scales nvt-hessian
+    # but a direct calculation may be considerd
+
+    nmol1, vol1 = nmolvol(state, nmolb, volumeb)
+    grad1 = similar(value(state))
+    grad1, hess = nvtgradienthessian!(grad1, hess, mix, nmol1, vol1, RT; buf=buf)
+
+    nmol2, vol2 = nmolb - nmol1, volumeb - vol1
+    grad2 = similar(value(state))
+    hess2 = similar(hess)
+    grad2, hess2 = nvtgradienthessian!(grad2, hess2, mix, nmol2, vol2, RT; buf=buf)
+
+    hess .+= hess2  # now, `hess` = hessian of a' + hessian of a''
+
+    # dNdN part scaling
+    @inbounds for j in 1:size(hess, 2)-1, i in 1:size(hess, 1)-1
+        hess[i, j] *= sqrt(nmol1[i] * nmol2[i] * nmol1[j] * nmol2[j] / (nmolb[i] * nmolb[j]))
+        if i == j
+            hess[i, j] += (grad1[i] - grad2[i]) * (nmol2[i] - nmol1[i]) / (2*nmolb[i])
+        end
+    end
+
+    # dNdV part scaling
+    @inbounds for i in 1:size(hess, 1)-1
+        sumnmolb = sum(nmolb)
+        hess[i, end] *= volumeb * sqrt(nmol1[i]*nmol2[i] / (nmolb[i] * sumnmolb))
+    end
+    hess[end, 1:end-1] .= hess[1:end-1, end]
+
+    # dV² part
+    hess[end, end] *= volumeb^2 / sum(nmolb)
+
+    return hess
+end
+
+function __vt_flash_optim_closures(
+    state1::IdealIdentityState,
+    mix::BrusilovskyEoSMixture,
+    nmolb::AbstractVector,
+    volumeb::Real,
+    RT::Real,
+)
+    state1x = value(state1)
+    buf = thermo_buffer(mix)
+
+    helmb = helmholtz(mix, nmolb, volumeb, RT; buf=buf)
+
+    nmol2 = similar(nmolb, Float64)
+
+    # Upper limits for state variables from definition 0 < x[i] < u[i]
+    # Necessary for constrain_step
+    xupperfromdef = [π * sqrt.(nmolb); sqrt(sum(nmolb))]
+
+    # covolumes = [nmolb .* components(mix).b; -volumeb]
+
+    function clsr_constrain_step(x::AbstractVector, dir::AbstractVector)
+        state1x .= x
+        α = Inf
+
+        # Finding upper limit
+        ## non-negativness: 0 < xi + α di < 1
+        for (i, (xi, di, ui)) in enumerate(zip(state1x, dir, xupperfromdef))
+            if iszero(di) && !(0 < xi < ui)
+                throw(ConstrainStepZeroDirectionError(i, xi))
+            end
+            α = di < 0 ? min(α, -xi/di) : min(α, (ui - xi)/di)
+        end
+
+        # ## covolumes
+        # state1xdotcov = dot(state1x, covolumes)
+        # dirdotcov = dot(dir, covolumes)
+        # if dirdotcov > 0
+        #     α = min(α, -state1xdotcov/dirdotcov)
+        # elseif dirdotcov < 0
+        #     α = min(α, (sum(covolumes) - state1xdotcov)/dirdotcov)
+        # end
+
+        # Finding lower limit
```

Comment:
Currently, for IdealStateVariables only non-negativness is considered, because covolume inequality is nonlinear.